### ISSUE TITLE
multiple test failures: stunnel fails to connect to backend: Connection refused

### ISSUE THREAD

**Issue by samueloph** [curl/curl#10682]
### I did this
\<a few commands that are part of building the Debian packages\>
...
`./runtests.pl -a -p !flaky -n`

https://buildd.debian.org/status/fetch.php?pkg=curl&arch=amd64&ver=7.88.1-2&stamp=1677835010&raw=0
https://buildd.debian.org/status/fetch.php?pkg=curl&arch=arm64&ver=7.88.1-2&stamp=1677835482&raw=0

### I expected the following

Successful execution of tests: 300 301 303 304 306 309 310 325 364 400 401 403 406 407 408 409 410 414 417 560 678 1112 1272 1561 1562 1630 1631 1632 2034 2037 2041 3000 3001 

### curl/libcurl version

curl 7.88.1 (x86_64-pc-linux-gnu) libcurl/7.88.1 OpenSSL/3.0.8 zlib/1.2.13 brotli/1.0.9 zstd/1.5.4 libidn2/2.3.3 libpsl/0.21.2 (+libidn2/2.3.3) libssh2/1.10.0 nghttp2/1.52.0 librtmp/2.3
Release-Date: 2023-02-20
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd

### operating system

Linux HOSTNAME 6.1.0-5-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.12-1 (2023-02-15) x86_64 GNU/Linux

### issue

Tests which make use of stunnel are currently flaky, but probably not flaky enough as this seems to be the first report about it.

Copying a snippet from Debian's bug report:
https://bugs.debian.org/1032343
```
> TESTFAIL: These test cases failed: 300 301 303 304 306 309 310 325 364 400 401 403 406 407 408 409 410 414 417 560 678 1112 1272 1561 1562 1630 1631 1632 2034 2037 2041 3000 3001

I've given these builds back in the hope that it's only an intermittent
failure. If I'm understanding the failure correctly, the test suite is
running stunnel to provide SSL/TLS in front of a backend server, but the
test is connecting to stunnel before the backend is ready, leading to
"Connection refused" errors:

> === Start of file https_stunnel.log
>  2023.03.03 09:09:04 LOG5[ui]: stunnel 5.68 on x86_64-pc-linux-gnu platform
>  2023.03.03 09:09:04 LOG5[ui]: Compiled/running with OpenSSL 3.0.8 7 Feb 2023
>  2023.03.03 09:09:04 LOG5[ui]: Threading:PTHREAD Sockets:POLL,IPv6,SYSTEMD TLS:ENGINE,OCSP,PSK,SNI Auth:LIBWRAP
>  2023.03.03 09:09:04 LOG5[ui]: Reading configuration from file /<<PKGBUILDDIR>>/debian/build/tests/https_stunnel.conf
>  2023.03.03 09:09:04 LOG5[ui]: UTF-8 byte order mark not detected
>  2023.03.03 09:09:04 LOG5[ui]: FIPS mode disabled
>  2023.03.03 09:09:04 LOG5[ui]: Configuration successful
>  2023.03.03 09:09:05 LOG5[0]: Service [curltest] accepted connection from ::ffff:127.0.0.1:51722
>  2023.03.03 09:09:05 LOG3[0]: s_connect: connect ::1:34321: Connection refused (111)
>  2023.03.03 09:09:05 LOG3[0]: No more addresses to connect
>  2023.03.03 09:09:05 LOG5[0]: Connection reset: 0 byte(s) sent to TLS, 0 byte(s) sent to socket
> === End of file https_stunnel.log
> === Start of file stderr300
>    % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                   Dload  Upload   Total   Spent    Left  Speed
> 
>    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>    0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
>  curl: (56) Recv failure: Connection reset by peer
> === End of file stderr300
```

I figured I should report this as it's also impacting Fedora. They workaround it by not installing stunnel on `aarch64` and `s390x` (apparently they didn't have this issue on `amd64`, like we did):
https://src.fedoraproject.org/rpms/curl/blob/d5c1163ef3a64c125452ce067670886765ecc5be/f/curl.spec#_104-108

I can confirm the same tests will sometimes pass on the same archs listed. 
This has come to my attention only recently due to improvements I'm doing to curl's packaging on Debian, which includes failing the build on failing tests.

cc'ing @kdudka (the Fedora maintainer) as I assume he'll be interested in the outcome of this too.

**Comment by bagder** [comment]
> but the test is connecting to stunnel before the backend is ready, leading to "Connection refused" errors:

I'm not convinced that's the reason. First: the logs show HTTP tests already running before test 300 so the HTTP server seems to be up and listening already. Then: when both the HTTP and HTTPS servers a fired up, the only way to do this is when the HTTP server has reported its port number back that it listens to and that means that it started *and has been verified to work*.

We run the test suite in numerous CI jobs in every PR, we can't recall having seen this problem there.

**Comment by kdudka** [comment]
@samueloph Right, we disable `stunnel` on `aarch64` and `s390x`: https://src.fedoraproject.org/rpms/curl/blob/d5c1163ef3a64c125452ce067670886765ecc5be/f/curl.spec#_104

We also disable a few unreliable test cases for specific architectures: https://src.fedoraproject.org/rpms/curl/blob/d5c1163ef3a64c125452ce067670886765ecc5be/f/curl.spec#_213

This is all needed to get reasonable success rate for Fedora curl builds.  Nevertheless I do not remember such a big list of tests that would fail during a single build.  It is usually 1-3 test cases what fails during the build on the less common arches.

**Comment by smcv** [comment]
I believe Debian's build servers that are described as "IPv6-only" have the loopback address `127.0.0.1` bound to `lo`, but no other IPv4 addresses (not even NAT).

One common reason for things to fail on machines configured like this is that if a component (not necessarily curl itself, maybe a piece of test infrastructure like stunnel) calls `getaddrinfo(..., "localhost", ...)` with the `AI_ADDRCONFIG` hint, `localhost` will not resolve to `127.0.0.1` as expected, because `AI_ADDRCONFIG` prevents that.

**Comment by bagder** [comment]
I bet the test suite would need some attention to work perfectly on a IPv6-only system.

**Comment by bagder** [comment]
While this might be a bug in the curl test suite, it is not high on my person agenda to make IPv6-only work. I'm more inclined to document IPv4 support as a requirement for the test suite.

**Comment by samueloph** [comment]
The evidence is strong on this being caused by IPv6-only hosts, so it's safe to assume this is the issue.

On Debian's side I have changed the packaging to ignore those test results, and we'll see if we manage to find exactly where the issue is.

Please feel free to do whatever you think it's best with this issue, whether you'd like to close it or use it to keep track of support for IPv-6 only test environments.

Thank you

**Comment by bagder** [comment]
No action on this.
